src/Microsoft.AspNet.SignalR.Core/Messaging/Subscription.cs

```diff
@@ -134,6 +134,7 @@ public async Task Work()
                 {
                     var messageResult = new MessageResult(items, totalCount);
 
+                    // REVIEW: If 'Invoke' faults, we fail to dispose and leak
```

Comment:
Hmmmm, file a bug?

Reply:
Yes, #4126 is that bug, and this is the fix.

I'm not 100% sure it's correct to catch the exception here but we could move it to there.

Reply:
Given that Invoke isn't ever supposed to throw (It would get logged as an error in  MessageBroker.DoWork), do you think it best to defensively catch any exceptions and dispose the subscription?

Reply:
Yeah, I wondered about that.

Reply:
It's harder to log the error from here since we don't have an `ITraceManager` and adding one would be tricky. That's the main reason I left it to the callback to handle the exception.

Reply:
I think calling `context.Subscription.Dispose()` in MessageBroker.DoWork's catch block is probably the way to go then. You'll have to make ISubscription implement IDisposable for that to work.

Reply:
Yeah, I like that better. I can test it a bit there too.